File: src/main/java/com/coralblocks/coralbench/Bench.java
/* 
 * Copyright 2015-2024 (c) CoralBlocks LLC - http://www.coralblocks.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
package com.coralblocks.coralbench;

import java.text.DecimalFormat;
import java.text.NumberFormat;
import java.util.Iterator;
import java.util.Map;
import java.util.TreeMap;

import com.coralblocks.coralbench.util.MutableInt;
import com.coralblocks.coralds.list.LinkedList;
import com.coralblocks.coralds.map.LongMap;

/**
 * <p>A class to benchmark Java code for latency. It prints some pretty results with percentiles. See an example below:</p>
 * <pre>
 * Measurements: 2,000,000 | Warm-Up: 1,000,000 | Iterations: 3,000,000
 * Avg Time: 1.025 micros | Min Time: 1.000 micro | Max Time: 56.500 micros
 * 75% = [avg: 1.016 micros, max: 1.042 micros]
 * 90% = [avg: 1.020 micros, max: 1.042 micros]
 * 99% = [avg: 1.022 micros, max: 1.042 micros]
 * 99.9% = [avg: 1.022 micros, max: 1.125 micros]
 * 99.99% = [avg: 1.023 micros, max: 7.791 micros]
 * 99.999% = [avg: 1.024 micros, max: 15.959 micros]
 * </pre> 
 */
public class Bench {

	private static final int DEFAULT_WARMUP = 0;
	private static final int NUMBER_OF_DECIMALS = 3;
	private static final String VERBOSE = "\n==CoralBenchVerbose==> ";
	private static final int DEFAULT_LOG_EVERY = 10000;

	private final DecimalFormat formatter = new DecimalFormat("#,###");
	
	private long time;
	private int iterations;
	private int measurements;
	private int toWarmup;
	private long totalTime;
	private long minTime;
	private long maxTime;
	
	private final LinkedList<MutableInt> pool = new LinkedList<MutableInt>(1024);
	private final LongMap<MutableInt> results = new LongMap<MutableInt>(4194304); // 2 ^ 22
	private final boolean verbose;
	private final int verboseLogEvery;
	
	/**
	 * Creates a <code>Bench</code> object without any warmup (warmup count is assumed to be zero)
	 */
	public Bench() {
		this(DEFAULT_WARMUP);
	}

	/**
	 * Creates a <code>Bench</code> object with the given warmup count
	 * 
	 * @param warmup how many iterations to warmup and ignore results
	 */
	public Bench(final int warmup) {
		
		this.toWarmup = warmup;
		
		try {
			// initialize it here so when you measure for the first time garbage is not created...
			Class.forName("java.lang.Math");
		} catch (final Exception e) {
			throw new RuntimeException(e);
		}
		
		String s = System.getProperty("coralBenchVerbose");
		this.verbose = s != null && s.equalsIgnoreCase("true");
		
		s = System.getProperty("coralBenchVerboseLogEvery");
		if (s == null) {
			this.verboseLogEvery = DEFAULT_LOG_EVERY;
		} else {
			this.verboseLogEvery = Integer.parseInt(s);
		}
		
		reset(true);
	}
	
	private final MutableInt getMutableInt(int x) {
		MutableInt mi = pool.removeLast();
		if (mi != null) {
			mi.set(x);
			return mi;
		}
		return new MutableInt(x);
	}
	
	private final void releaseMutableInt(MutableInt mi) {
		pool.addLast(mi);
	}
	
	/**
	 * Reset this benchmark so that it can be re-used for a new and fresh benchmark. If this benchmark has a warmup count, it will NOT be repeated.
	 */
	public void reset() {
		reset(false);
	}

	/**
	 * Reset this benchmark so that it can be re-used for a new and fresh benchmark.
	 * 
	 * @param repeatWarmup should warmup be repeated or not?
	 */
	public void reset(boolean repeatWarmup) {
		time = 0;
		iterations = 0;
		totalTime = 0;
		if (!repeatWarmup) toWarmup = 0;
		minTime = Long.MAX_VALUE;
		maxTime = Long.MIN_VALUE;
		
		measurements = 0;
		Iterator<MutableInt> iter = results.iterator();
		while(iter.hasNext()) {
			releaseMutableInt(iter.next());
		}
		results.clear();
	}
	
	/**
	 * Starts the timer to measure
	 * 
	 * @return the start timestamp in nanoseconds
	 */
	public final long mark() {
		time = System.nanoTime();
		return time;
	}
	
	/**
	 * Measure (and possibly store if not warming up) the elapsed time
	 * 
	 * @return the elapsed time in nanoseconds
	 */
	public long measure() {
		if (time > 0) {
			long now = System.nanoTime();
			long elapsed = now - time;
			if (elapsed < 0) throw new RuntimeException("Found a negative elapsed time: now=" + now + " start=" + time + " diff=" + elapsed);
			final boolean counted = measure(elapsed);
			if (counted) {
				return elapsed;
			}
		}
		return -1;
	}
	
	/**
	 * Are we currently warming up or are we currently measuring?
	 * 
	 * @return true if warming up
	 */
	public final boolean isWarmingUp() {
		return toWarmup <= iterations;
	}

	/**
	 * Register and take into account (and possibly store if not warming up) the provided elapsed time
	 * 
	 * @param lastNanoTime the elapsed time you want to take into account
	 * @return true if it was taken into account because we are not warming up
	 */
	public final boolean measure(final long lastNanoTime) {
		
		final boolean isToMeasure = ++iterations > toWarmup;
		
		if (isToMeasure) {

			totalTime += lastNanoTime;
			minTime = Math.min(minTime, lastNanoTime);
			maxTime = Math.max(maxTime, lastNanoTime);
			
			MutableInt count = results.get(lastNanoTime);
			if (count == null) {
				results.put(lastNanoTime, getMutableInt(1));
			} else {
				count.set(count.get() + 1);
			}
			measurements++;
		}
		
		if (verbose) {
			if (iterations == 1 || iterations % verboseLogEvery == 0) System.out.println(VERBOSE + "Iteration " + iterations + " => " + lastNanoTime); 
		}
		
		return isToMeasure;
	}
	
	private String formatPercentage(double x, int decimals) {
		NumberFormat percentFormat = NumberFormat.getPercentInstance();
		percentFormat.setMaximumFractionDigits(decimals);
		return percentFormat.format(x);
	}
	
	private void addPercentile(StringBuilder sb, double perc, TreeMap<Long, MutableInt> treeMap) {
		
		if (treeMap.isEmpty()) {
			return;
		}

		long maxTop = -1;
		
		long x = Math.round(perc * measurements);
		Iterator<Map.Entry<Long, MutableInt>> iter = treeMap.entrySet().iterator();
		int iTop = 0;
		long sumTop = 0;
		
		LOOP: while(iter.hasNext()) {
			
			Map.Entry<Long, MutableInt> entry = iter.next();
			Long time = entry.getKey();
			MutableInt count = entry.getValue();
			
			for(int a = 0; a < count.get(); a++) {
				
				iTop++;
				sumTop += time;
				
				if (iTop == x) {
					maxTop = time;
					break LOOP;
				}
			}
		}
		
		sb.append(formatPercentage(perc, 8));
		sb.append(" = [");
		sb.append("avg: ").append(convertNanoTime(sumTop / iTop));
		sb.append(", max: ").append(convertNanoTime(maxTop)).append(']');
		sb.append('\n');
	}
	
	/**
	 * How many iterations have we done so far, including warming up?
	 * 
	 * @return the total number of iterations so far
	 */
    public int getIterations() {
    	return iterations;
    }
    
    /**
     * How many measurements have we done so far, excluding warming up?
     * 
     * @return the total number of measurements so far
     */
    public int getMeasurements() {
    	return measurements;
    }
	
    /**
     * Return the average latency of all measurements
     * 
     * @return the average latency
     */
	public final double getAverage() {
		return avg();
	}

	private final double avg() {
		if (measurements <= 0) {
			return 0;
		}
		final double avg = ((double) totalTime / (double) measurements);
		final double rounded = Math.round(avg * 100D) / 100D;
		return rounded;
	}
	
	private static double round(double d) {
		return round(d, NUMBER_OF_DECIMALS);
	}
	
	private static double round(double d, int decimals) {
		double pow = Math.pow(10, decimals);
		return ((double) Math.round(d * pow)) / pow;
	}
	
	private static String convertNanoTime(double nanoTime) {
		StringBuilder sb = new StringBuilder();
		convertNanoTime(nanoTime, sb);
		return sb.toString();
	}
	
	private static StringBuilder convertNanoTime(double nanoTime, StringBuilder sb) {
		if (nanoTime >= 1000000000L) {
			// seconds...
			double seconds = round(nanoTime / 1000000000D);
			sb.append(formatThreeDecimals(seconds)).append(seconds > 1 ? " secs" : " sec");
		} else if (nanoTime >= 1000000L) {
			// millis...
			double millis = round(nanoTime / 1000000D);
			sb.append(formatThreeDecimals(millis)).append(millis > 1 ? " millis" : " milli");
		} else if (nanoTime >= 1000L) {
			// micros...
			double micros = round(nanoTime / 1000D);
			sb.append(formatThreeDecimals(micros)).append(micros > 1 ? " micros" : " micro");
		} else {
			double nanos = round(nanoTime);
			sb.append(formatThreeDecimals(nanos)).append(nanos > 1 ? " nanos" : " nano");
		}
		return sb;
	}
	
	private static String formatThreeDecimals(double d) {
		return String.format("%.3f", d);
	}

	/**
	 * Get the results as a string
	 * 
	 * @param includePercentiles true to include percentiles
	 * @return the results as a string
	 */
	public String results(boolean includePercentiles) {
		final StringBuilder sb = new StringBuilder(128);
		sb.append("Measurements: ").append(formatter.format(measurements));
		sb.append(" | Warm-Up: ").append(formatter.format(toWarmup));
		sb.append(" | Iterations: ").append(formatter.format(iterations)).append('\n');
		if (measurements > 0) {
			sb.append("Avg Time: ").append(convertNanoTime(avg()));
			sb.append(" | Min Time: ").append(convertNanoTime(minTime));
			sb.append(" | Max Time: ").append(convertNanoTime(maxTime));

			sb.append('\n');
		
			if (includePercentiles) {
				TreeMap<Long, MutableInt> treeMap = new TreeMap<Long, MutableInt>();
				Iterator<MutableInt> iter = results.iterator();
				while(iter.hasNext()) {
					MutableInt counter = iter.next();
					Long time = results.getCurrIteratorKey();
					treeMap.put(time, counter);
				}
		
				addPercentile(sb, 0.75D, treeMap);
				addPercentile(sb, 0.9D, treeMap);
				addPercentile(sb, 0.99D, treeMap);
				addPercentile(sb, 0.999D, treeMap);
				addPercentile(sb, 0.9999D, treeMap);
				addPercentile(sb, 0.99999D, treeMap);
			}
		}
		
		return sb.toString();
	}
	
	/**
	 * This method simply calls <code>results(true)</code>.
	 * 
	 *  @return the results as a string
	 */
	public String results() {
		return results(true);
	}
	
	/**
	 * Print the results to stdout
	 * 
	 * @param includePercentiles true to include percentiles
	 */
	public void printResults(boolean includePercentiles) {
		System.out.println(results(includePercentiles));
	}
	
	/**
	 * This method simply calls <code>printResults(true)</code>.
	 */
	public void printResults() {
		printResults(true);
	}
	
}

File: src/main/c/bench.cpp
/* 
 * Copyright 2024 (c) CoralBlocks - http://www.coralblocks.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
#include "bench.hpp"
using namespace std;

Bench::Bench(int warmupCount)
    : warmupCount(warmupCount),
      measurementCount(0),
      sum(0),
      minTime(numeric_limits<long long>::max()),
      maxTime(numeric_limits<long long>::min()),
      size(0) {

        results = new map<long long, long long>();

}

Bench::~Bench() {
    delete results;
}

void Bench::mark() {
    startTime = chrono::steady_clock::now();
}

void Bench::measure() {
    auto endTime = chrono::steady_clock::now();
    auto elapsed = chrono::duration_cast<chrono::nanoseconds>(endTime - startTime).count();
    measure(elapsed);
}

bool Bench::measure(long long elapsed) {

    bool isToMeasure = ++measurementCount > warmupCount;

    if (isToMeasure) {
        sum += elapsed;
        if (elapsed < minTime) minTime = elapsed;
        if (elapsed > maxTime) maxTime = elapsed;

        // Increment the frequency of this elapsed time
        auto it = results->find(elapsed);
        if (it == results->end()) {
            results->insert({elapsed, 1});
        } else {
            it->second++;
        }
        size++;
    }
    
    return isToMeasure;
}

int Bench::getIterations() const {
	return measurementCount;
}

int Bench::getMeasurements() const {
    return size;
}

void Bench::reset() {
	reset(false);
}

void Bench::reset(bool repeatWarmup) {
    measurementCount = 0;
    sum = 0;
    if (!repeatWarmup) warmupCount = 0;
    minTime = numeric_limits<long long>::max();
    maxTime = numeric_limits<long long>::min();
    results->clear();
    size = 0;
}

bool Bench::isWarmingUp() const {
	return warmupCount <= measurementCount;
}

double Bench::avg() const {
    const int effectiveCount = measurementCount - warmupCount;
    if (effectiveCount <= 0) {
        return 0;
    }
    const double avg = static_cast<double>(sum) / effectiveCount;
    const double rounded = round(avg * 100.0) / 100.0;
    return rounded;
}

double Bench::getAverage() const {
	return avg();
}	

void Bench::printResults() const {
	printResults(true);
}

void Bench::printResults(bool includePercentiles) const {

    int effectiveCount = measurementCount - warmupCount;

    string effCountStr = formatWithCommas(effectiveCount);
    string warmupStr = formatWithCommas(warmupCount);
    string totalStr = formatWithCommas(measurementCount);

    cout << "Measurements: " << effCountStr
         << " | Warm-Up: " << warmupStr
         << " | Iterations: " << totalStr << endl;
         
  	if (effectiveCount > 0) {

	    auto [avgVal, avgUnit] = formatTime(avg());
	    auto [minVal, minUnit] = formatTime(static_cast<double>(minTime));
	    auto [maxVal, maxUnit] = formatTime(static_cast<double>(maxTime));
	
	    cout << fixed << setprecision(3);
	    cout << "Avg Time: " << avgVal << " " << avgUnit << " | "
	         << "Min Time: " << minVal << " " << minUnit << " | "
	         << "Max Time: " << maxVal << " " << maxUnit << endl;
	
	    if (includePercentiles) printPercentiles();
	}
	
	cout << endl;
}

string Bench::formatWithCommas(long long value) {
    string numStr = to_string(value);
    int insertPosition = static_cast<int>(numStr.length()) - 3;
    while (insertPosition > 0) {
        numStr.insert(insertPosition, ",");
        insertPosition -= 3;
    }
    return numStr;
}

pair<double, string> Bench::formatTime(double nanos) {
    if (nanos >= 1'000'000'000.0) {
        double seconds = nanos / 1'000'000'000.0;
        return {roundToDecimals(seconds, 3), seconds > 1 ? "seconds" : "second"};
    } else if (nanos >= 1'000'000.0) {
        double millis = nanos / 1'000'000.0;
        return {roundToDecimals(millis, 3), millis > 1 ? "millis" : "milli"};
    } else if (nanos >= 1000.0) {
        double micros = nanos / 1000.0;
        return {roundToDecimals(micros, 3), micros > 1 ? "micros" : "micro"};
    } else {
        double ns = nanos;
        return {roundToDecimals(ns, 3), ns > 1 ? "nanos" : "nano"};
    }
}

double Bench::roundToDecimals(double d, int decimals) {
    double pow10 = pow(10.0, decimals);
    return round(d * pow10) / pow10;
}

void Bench::printPercentiles() const {

    if (size == 0) return;

    double percentiles[] = {0.75, 0.90, 0.99, 0.999, 0.9999, 0.99999};

    for (double p : percentiles) {
        addPercentile(p);
    }
}

string Bench::formatPercentage(double perc) {
    double p = perc * 100.0;

    ostringstream oss;
    oss << fixed << setprecision(6) << p;

    string s = oss.str();
    // remove trailing zeros
    while (s.back() == '0') {
        s.pop_back();
    }

    // if the last character is now a '.', remove it
    if (s.back() == '.') {
        s.pop_back();
    }

    // Append the '%' sign
    s += "%";

    return s;
}

void Bench::addPercentile(double perc) const {

    if (results->empty()) return;

    long long target = static_cast<long long>(llround(perc * size));
    if (target == 0) return;
    if (target > size) target = size;

    // Iterate through the map to find the top element at position target
    long long iTop = 0;
    long long sumTop = 0;
    long long maxTop = -1;

    for (auto &entry : *results) {
        long long time = entry.first;
        long long count = entry.second;

        for (int i = 0; i < count; i++) {
            iTop++;
            sumTop += time;
            if (iTop == target) {
                maxTop = time;
                goto FOUND;
            }
        }
    }

FOUND:;

    double avgTop = static_cast<double>(sumTop) / iTop;
    auto [avgVal, avgUnit] = formatTime(avgTop);
    auto [maxVal, maxUnit] = formatTime(static_cast<double>(maxTop));

    cout << fixed << setprecision(3);
    cout << formatPercentage(perc) << " = [avg: " << avgVal << " " << avgUnit
         << ", max: " << maxVal << " " << maxUnit << "]\n";
}
